mp4/avcdecoderconfigurationrecord.go

```diff
@@ -75,26 +75,58 @@ func DecodeAVCDecConfRec(r io.Reader) (AVCDecConfRec, error) {
 }
 
 // Encode - write an AVCDecConfRec to w
-func (a *AVCDecConfRec) Encode(w io.Writer) {
+func (a *AVCDecConfRec) Encode(w io.Writer) error {
+	var err error
+	write := func(b byte) {
+		if err != nil {
+			return
+		}
+		err = binary.Write(w, binary.BigEndian, b)
+	}
+
 	var configurationVersion byte = 1
 	var ffByte byte = 0xff
-	binary.Write(w, binary.BigEndian, configurationVersion)
-	binary.Write(w, binary.BigEndian, a.AVCProfileIndication)
-	binary.Write(w, binary.BigEndian, a.ProfileCompatibility)
-	binary.Write(w, binary.BigEndian, a.AVCLevelIndication)
-	binary.Write(w, binary.BigEndian, ffByte)     // Set length to 4
+	write(configurationVersion)
+	write(a.AVCProfileIndication)
+	write(a.ProfileCompatibility)
+	write(a.AVCLevelIndication)
+	write(ffByte) // Set length to 4
+	if err != nil {
+		return err
+	}
```

Comment:
This could probably be done nicer, but didn't want to spend too much time on this. This approach is taken from https://blog.golang.org/errors-are-values.

Reply:
OK. Thanks for the reference. Having an if statement after each write would be extremely ugly, so this is definitely much better although the side effect make it somewhat hard to understand.

Reply:
The code somewhat further done in the blog by Rob Pike you referred to has a nicer approach with a struct. Will probably try to change to that later.

Reply:
The `struct` approach is indeed nicer. But the problem is it mostly works for `[]byte` instead of a mix of `byte` and `unint16`. So you need multiple implementations. It also works better for the generic `Write()` and not `binary.Write`. But I think I have a composable solution for that:

```go
type errWriter struct {
	w io.Writer
	err error
}

func (ew *errWriter) write(buf []byte) {
	if ew.err != nil {
		return
	}
	_, ew.err = ew.w.Write(buf)
}

type bigEndianWriter struct {
	w io.Writer
}

func (bew *bigEndianWriter) Write(p []byte) (int, error) {
	return len(p), binary.Write(bew.w, binary.BigEndian, p)
}
```

Then you can use it like this:

```go
bew := &bigEndianWriter{w: fd}

ew := errWriter{w: bew}

ew.write(p0[a:b])
ew.write(p1[c:d])
ew.write(p2[e:f])

if ew.err != nil {
	return ew.err
}
```

It's honestly a bit unfortunate that `binary.Write` doesn't implement the `io.Writer` interface.

Reply:
Thanks. binary.Write() has `data interface{}` as its third argument and writes as many bytes as the argument requires, eg. 4 bytes for uint32. Therefore, I think it makes more sense to give errWriter a method `write(data interface{} )` that can handle the same data as `bytes.Writer()`.